auth/policy.go

```diff
@@ -0,0 +1,143 @@
+package auth
```

Comment:
A really nice abstraction, and if this wasn't an open source project, I would say definitely keep it here. However, I think we should move this into our own lambda authoriser implementation, and keep g8 as a really thin and dumb layer.

We should allow the users of this library to handle the AWS SDK response object themselves directly. I think g8 in general is quite a dumb layer in an application, and has quite a light touch on how people implement their apps. The benefit it gives us is really about new relic integration, logging, and simpler interface than the default one given by AWS SDK.

Reply:
Having thought about this some more, I think it doesn’t need to be moved, but it needs some changes to be more inline with how other handlers work in g8. Instead of having a custom response struct, I think we should have the sdk response struct in the context, and then we can just make some methods on the context which modify the sdk response object.

Reply:
Hope that makes sense - shout if you have any questions.

Reply:
Hope that makes sense - shout if you have any questions.

Reply:
I think we want to avoid having the custom response object in the public API of the context. I would want the fields to look like this:

```
type APIGatewayCustomAuthorizerContext struct {
	Context       context.Context
	Request       events.APIGatewayCustomAuthorizerRequestTypeRequest
	Response      events.APIGatewayCustomAuthorizerRequestTypeResponse
	Logger        zerolog.Logger
	NewRelicTx    newrelic.Transaction
	CorrelationID string
}
```

but you should replace your custom struct to just methods on the context. For example, you should implement:

```
func (r *APIGatewayCustomAuthorizerContext) SetPrincipalID(principalID string) {
	...
}

func (r *APIGatewayCustomAuthorizerContext) AllowAllMethods() {
	...
}

func (r *APIGatewayCustomAuthorizerContext) DenyAllMethods() {
	...
}

func (r *APIGatewayCustomAuthorizerContext) AllowMethod(verb HttpVerb, resource string) {
	...
}

func (r *APIGatewayCustomAuthorizerContext) DenyMethod(verb HttpVerb, resource string) {
	...
}

func (r *APIGatewayCustomAuthorizerContext) HasAllowingMethod() bool {
  ...
}
```

and you should either:
* directly manipulate the sdk response object inside these implementations
* or maybe change some private attribute of the context to store them